### PR TITLE
[WIP] Retinanet Training II

### PR DESCRIPTION
**Pull request by mszsorondo**
#1636

next (high level) steps: 

- [x] L1 loss
- [x] focal loss for anchor classifications
- [x] L1 loss for box regressions
- [x] clean up code
- [x] adapt for mlperf-retinanet training specs
- [x] refactor for exploiting tensor lazyness and tensor mul + JIT
- [x] extra spec details (e.g. img flipping, random sampling, etc)
- [ ] Final refactor

### CODE REVIEW

**Comment by mszsorondo** [comment]
@geohot I think this is almost done, my laptop doesn't have much resources to run big enough training samples to check (hw, if available, is 3x more expensive in argentina). It already takes me too long with resized images of 200x200. Can you SSH me to some machine? not necesarily the AMDs yet. Just something decent. I've been checking our implementation with mlperf's by calling for their constructors from tinygrad instead of running the actual mlperf's official training scripts. I am probably skipping some details I'd like to see by watching the step-by-step of their scripts with the original batch and image sizes.

**Comment by geohot** [comment]
Do you have an estimated time to completion? How much will you wager it's correct?

**Comment by mszsorondo** [comment]
> Do you have an estimated time to completion? How much will you wager it's correct?

I'm 100% confident this reproduces what the official [**written** spec](https://github.com/mlcommons/training/tree/master/single_stage_detector#datasetenvironment) states.

But there may be some missing details that cannot be seen unless you run the actual code from their reference code. If there is something missing, it should not be difficult to solve. I'll push one more commit that sets things up for running our implementation with the full training set and original batch & image sizes. This is why I don't want to ask for the AMDs now.

3 day worst-case time to completion once I have access to a computer that lets me reproduce a couple iterations from the training loop for both implementations. I'll try to get one elsewhere if this isn't possible or generates some trouble.

**Comment by gijskoning** [comment]
If you want to be more confident of the runtime of your code on a bigger machine I recommend vast.ai. Its very cheap (0.24 cents/hour for rtx 3090) and easy to setup. They only list nvidia gpus but that should be enough to test your code.

**Comment by geohot** [comment]
Unlocking bounty, over 1 week has gone by without update.

**Comment by geohot** [comment]
New rule on mlperf training bounties. in order to lock the bounty, you must post weights that you trained in tinygrad that meet the accuracy requirement. Then we can look at speed and mergablity, but until there's weights, no lock.
